File: logmine/LogMine.py
from argparse import ArgumentParser
from boom.common.utils import init_logger
from boom.common.utils import dynamite_parser
from boom.dynamite.DynamitePod import DynamitePod
from datetime import datetime, timedelta
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from importlib import import_module
import re
import smtplib
import traceback

FILTER2 = True

if FILTER2:
    from FilterFinder import FilterFinder

SEVERITY = {
               0:  "DEFAULT",    # (0) The log entry has no assigned severity level",
              100: "DEBUG",      # (100) Debug or trace information",
              200: "INFO",       # (200) Routine information, such as ongoing status or performance",
              300: "NOTICE",     # (300) Normal but significant events, such asa configuration change",
              400: "WARNING",    # (400) Warning events might cause problems",
              500: "ERROR",      # (500) Error events are likely to cause problems",
              600: "CRITICAL",   # (600) Critical events cause more severe problems or outages",
              700: "ALERT",      # (700) A person must take an action immediately",
              800: "EMERGENCY",  # (800) One or more systems are unusable"
}


class LogMine():
    def __init__(self):
        self.logger = init_logger('LogMine')
        self.args = self.arguments()
        self.results = list()
        self.new_data = dict()
        self.stopfile_data = {}
        self.filterfile_data = []
        self.min_len = 16
        self.max_len = 64
        self.max_substring_length = 128  # truncate log entry payloads to 1st N characters if FILTER2
        self.df = None  # pandas dataframe

        self.dynamite = DynamitePod(pod=self.args.pod,
                                    realm=self.args.realm,
                                    zone=self.args.zone,
                                    project=self.args.project,
                                    cloud=self.args.cloud,
                                    user=self.args.user)

        if self.args.send_file:
            self.send_requested_file()
        else:
            self.logger.info('do query')
            entries = self.do_query()
            self.logger.debug('list entries')
            for entry in entries:
                self.logger.debug(entry)
            self.logger.info('output raw results to file')
            self.write_raw_results('{}/{}'.format(self.args.output_dir, self.args.output_file), entries)

            # look for new input (based on what is not in the stopfile) and report it, then
            # update the stopfile with the new input
            # use optional output_file arg in case of pre-existing file (no query and no parse)
            self.logger.info('do process')
            self.do_process(entries)

            # send notification emails if sender, recipients and new_data        if self.args.sender and self.args.recipients and self.new_data:
            if self.new_data:
                self.logger.info('send notification(s)')
                for recipient in self.args.recipients:
                    self.do_sendmail(self.args.sender, recipient, self.new_data)

            # write out the new filter_file, including existing and new entries found by FilterFinder
            self.logger.info('write filter file')
            self.write_filter_file()

    def arguments(self):
        """
        argument parser
        :return: parsed args
        """
        parser = ArgumentParser(parents=[dynamite_parser()])
        parser.add_argument('--limit', default=100, type=int, help='number of log entries to retrieve')
        parser.add_argument('--max_entry_count', default=20, type=int, help='max per log entry to save to file')
        parser.add_argument('--output_dir', default='/tmp')
        parser.add_argument('--range_in_minutes', type=int, default=0)
        parser.add_argument('--sender', default='no-reply@nestlabs.com')
        parser.add_argument('--recipients', nargs='+', default=['tlichtenberg@google.com'])
        parser.add_argument('--output_file',  required=True)
        parser.add_argument('--send_file', default=False, type=bool, help="request an email containing the file specified in output_dir and output_file")
        known_args = parser.parse_known_args()

        # dependent args
        parser.add_argument('--log_filter', required=known_args[0].send_file == False)
        parser.add_argument('--source', required=known_args[0].send_file == False,
                            help="for stopfile_<source>.json and newfile_<source>.json")

        return parser.parse_args()

    def do_query(self):
        """
        performs the stackdriver log query, putting annotated results into the self.results list
        :return:
        """
        # initialize raw_entries, filterfile_data and optional filter revisions
        raw_entries = []
        entries = []
        self.read_filter_file()
        self.do_filter_revisions()
        self.logger.info(self.args.log_filter)

        # get log entries
        try:
            raw_entries = self.get_log_entries()
        except:
            self.logger.error(traceback.format_exc())

        # generate live append to filterfile_data from the raw_entries by analyzing for common substrings
        payloads = []
        for entry in raw_entries:
            self.logger.debug('* * * RAW ENTRY: {}'.format(entry))
            payload = self.get_payload(entry)
            final_entry = self.prep_entry(payload)
            self.logger.debug('* * * FINAL ENTRY: len() {}, {}'.format(len(final_entry), final_entry))
            payloads.append(final_entry)

        if FILTER2:
            p2 = []
            for p in payloads:
                p = p[:self.max_substring_length]  # truncate for substring matching
                p2.append(p)
            self.logger.debug('len payloads list: {}'.format(len(p2)))
            if len(p2) > 0:
                finder = FilterFinder(p2)
                finder_list = finder.run()
                for item in finder_list:
                    self.logger.info("adding {} to filterfile_data".format(item))
                    self.filterfile_data.append(item)

        for payload in payloads:
            output = [payload]
            if output:
                output[0] = self.do_filter(output[0])
                self.logger.debug('payload: {}'.format(output))
                output.insert(0, datetime.fromtimestamp(entry.timestamp.seconds).strftime('%Y-%m-%dT%H:%M:%SZ'))
                output.insert(0, SEVERITY[entry.severity])
                entries.append(str(output))

        return entries

    def prep_entry(self, entry):
        """
        same as in FilterFinder, strips unwanted chars and removes multiple white spaces within line
        :param entry: the log entry payload
        :return: sanitizing string
        """
        entry = entry.strip().replace('[', '').replace(']','').replace("'","").replace('"',"")
        entry = re.sub( '\s+', ' ', entry ).strip()
        self.logger.debug('len(item): {}, item: {}'.format(len(entry),entry))
        return entry

    def get_payload(self, entry):
        """
        in the python logging api, the payload can be of several types. in go it's always just the entry.payload
        :param entry: the raw log entry
        :return: the actual payload
        """
        payload = ''
        if len(entry.text_payload) > 0:
            try:
                payload = str(entry.text_payload).replace(',','').replace('\n', '') # lose any commas in the entry string
            except:
                self.logger.error(traceback.format_exc())

        elif entry.json_payload:
            try:
                payload = str(entry.json_payload['message']).replace(',', '').replace('\n', '')
            except:
                self.logger.error(traceback.format_exc())

        elif entry.proto_payload:
            class_ = getattr(import_module('google.protobuf.any_pb2'), 'Any')  # class reference
            try:
                if entry.proto_payload.type_url.find('AuditLog') < 0:
                    rv = class_()  # instantiate class instance
                    rv.ParseFromString(entry.proto_payload.value)  # decode Any.value
                    payload = str(rv)
                else:
                    payload = str(entry.proto_payload.value).replace('\n', '')
            except:
                self.logger.error(traceback.format_exc())

        return payload

    def do_filter(self, log_entry):
        """
        filterfile_data contains substrings of common log entries. if the given entry finds one of those
        substrings, replace the entry with that substring.
        for example:

        :param log_entry: a log entry
        :return:
        """
        for line in self.filterfile_data:
            line = line.rstrip('\n').strip()
            self.logger.debug('checking line {}'.format(line))
            if log_entry.find(line) >= 0:
                self.logger.debug('found line {}'.format(line))
                log_entry = line
                break
        return log_entry

    def do_filter_revisions(self):
        """
         optional filter revisions for timestamp and dataflow job name as applicable

         for dataflow jobs we need the job_id not the job name
         the caller should provide a dataflow job name as the --source arg
         for example:
        --source your-data-flow-job-name
         --log_filter 'resource.type=dataflow_step resource.labels.job_id=DATAFLOW_JOB_NAME'
        :return:
        """
        if self.args.log_filter.find('DATAFLOW_JOB_NAME') >= 0:
            job = self.dynamite.dataflow_client.get_job_by_name(self.args.source)
            if job:
                self.args.log_filter = self.args.log_filter.replace('DATAFLOW_JOB_NAME', job['id'])
            else:
                raise Exception('did not find active dataflow job for --source: {}'.format(self.args.source))

        if self.args.range_in_minutes > 0:
            start_time = (datetime.now() - timedelta(minutes=self.args.range_in_minutes)).strftime('%Y-%m-%dT%H:%M:%SZ').upper()
            self.args.log_filter += " timestamp > \"{}\"".format(start_time)

    def get_log_entries(self):
        """
        interface to google.cloud.logging client.list_entries
        """
        _, _, raw_entries = self.dynamite.logging.get_log_entries(log_filter=self.args.log_filter, limit=self.args.limit)
        self.logger.debug('len(raw_entries): {}'.format(len(raw_entries)))
        return raw_entries

    def write_raw_results(self, outfile=None, entries=None):
        """
        write to disk
        :param outfile: optional output file path. defaults to args.output_file
        :param entries: list of [SEVERITY, TIMESTAMP, TEXT] log entries
        :return:
        """
        if entries:
            if not outfile:
                outfile = self.args.output_file

            # get rid of hex code stuff if there is any and write to file
            outfile_list = outfile
            with open(outfile_list, 'w') as f:
                for r in entries:
                    r = str(re.sub(r"\\x[0-9a-fA-F]{2}", "", str(r)))
                    f.write('{}\n'.format(r))

    def do_process(self, entries):
        """
        :param entries: list of [SEVERITY, TIMESTAMP, TEXT] log entries

        process the log entries against the input stopfile_data
        discover new data and report it
        overwrite the stopfile and write the newfile to disk
        :return:
        """

        # read in stopfile
        self.read_stop_file()

        data_counts = {}

        # iterate through the log entry dataset
        # for this we are only interested in the TEXT portion of the log entries
        # and in that, we want the first part (the log entry, not its count list)
        for entry in entries:
            row = entry.split(',')
            txt = row[2].strip()
            items = txt.rsplit(":")
            item = self.prep_entry(items[0])
            if len(item) < self.min_len or len(item) > self.max_len:
                continue
            # add to counts of unique keys in this dataset
            if data_counts.has_key(item):
                data_counts[item] += 1
            else:
                data_counts[item] = 1

        # add new counts to existing stopfile entries
        # or create new stopfile entry if new data
        for k,v in data_counts.iteritems():
            if len(k) == 0:
                continue
            self.logger.debug('k = {}, v = [{}]'.format(k, v))
            if k in self.stopfile_data.keys():
                self.stopfile_data[k].append(v)
                if len(self.stopfile_data[k]) > self.args.max_entry_count:
                    del self.stopfile_data[k][:self.args.max_entry_count]  # keep the last max_entry_count elements only
            else:
                self.stopfile_data[k] = []
                self.stopfile_data[k].append(v)
                self.new_data[k] = [v]

        # write out new stopfile
        new_stopfile = []
        for k, v in self.stopfile_data.iteritems():
            new_stopfile.append("{}:{}".format(k,str(v)))
        self.write_stop_file(new_stopfile)

        # report new data and write out newfile
        new_newfile = []
        for k, v in self.new_data.iteritems():
            new_newfile.append("{}:{}".format(k,str(v)))
        self.write_new_file(new_newfile)

    def write_text_file(self, outfile, lines):
        """
        utility to write a list of lines to a text file
        :param outfile: path to the file
        :param lines: list of text lines
        :return:
        """
        f = open(outfile, 'w')
        for line in lines:
            f.write("{}\n".format(line))
        f.flush()
        f.close()

    def read_text_file(self, infile):
        """
        utility to read text file into a list
        :param infile: the file path
        :return: list of text lines
        """
        lines = []
        raw_lines = open(infile, 'r').readlines()
        for r in raw_lines:
            lines.append(r.replace('\n',''))
        return lines

    def write_stop_file(self, stopfile):
        """
        writes the stopfile entries to a text file
        :param stopfile: the stopfile entries
        :return:
        """
        outfile = '{}/stopfile_{}.txt'.format(self.args.output_dir, self.args.source)
        self.write_text_file(outfile, stopfile)

    def read_stop_file(self, infile=None):
        """
        read the stopfiles from disk and parse into the self.stopfiles_data dictionary
        :return:
        """
        try:
            if not infile:
                infile = '{}/stopfile_{}.txt'.format(self.args.output_dir, self.args.source)
            lines = self.read_text_file(infile)
            for l in lines:
                counts = []
                line = l.split(":")
                # line 0 is the log entry, line 1 is a str list of counts that needs to be turned back into an int list
                value = self.prep_entry(line[1])
                values = value.split(',')
                for v in values:
                    counts.append(int(v))
                self.stopfile_data[line[0]] = counts
        except:
            self.stopfile_data = {}

    def write_new_file(self, newfile):
        """
        writes the newfile entries to a text file
        :param newfile: the newfile entries
        :return:
        """
        outfile = '{}/newfile{}.txt'.format(self.args.output_dir, self.args.source)
        self.write_text_file(outfile, newfile)

    def read_filter_file(self, infile=None):
        """
        reads in an existing filterfile or initializes an empty list for that field
        """
        if not infile:
            infile = '{}/filterfile_{}.txt'.format(self.args.output_dir, self.args.source)
        try:
            self.filterfile_data = []
            with open(infile) as f:
                data = f.readlines()
                for d in data:
                    self.filterfile_data.append(d.rstrip('\n'))
        except:
            self.filterfile_data = []

    def write_filter_file(self):
        """
        writes out a new filterfile with existing and additional filters found
        """
        outfile = '{}/filterfile_{}.txt'.format(self.args.output_dir, self.args.source)
        self.write_text_file(outfile, self.filterfile_data)

    def send_requested_file(self):
        """
        read and format requested stop or filter files and send as email to recipients
        :return:
        """
        infile = '{}/{}'.format(self.args.output_dir, self.args.output_file)
        if infile.find('filterfile') >= 0:
            self.read_filter_file(infile)
            new_data = {}
            for item in self.filterfile_data:  # format into a dictionary for sendmail processing
                new_data[item] = 1
        elif infile.find('stopfile') >= 0:
            self.read_stop_file(infile)
            new_data = self.stopfile_data
        else:
            self.logger.error('only stopfiles and filterfiles are supported by send_file')
            return

        subject = "(%s) Log Mine: %s - %s" % (self.args.project, infile, datetime.now().strftime("%Y-%m-%d"))

        self.logger.info('send notification(s)')
        for recipient in self.args.recipients:
            self.do_sendmail(self.args.sender, recipient, new_data, subject)

    def do_sendmail(self, sender, recipient, new_data, subject=None):
        """
            general email routine to notify recipients of new entries
        """
        body = "<body><table border=2><tr><td> Entries </td><td> Count </td></tr>"
        for key in sorted(new_data.keys()):
            body_part = "<tr><td>%s</td><td>%s</td></tr>" % \
                        (key, str(new_data[key]))
            body += body_part
        body += "</table></body>"
        if not subject:
            subject = "(%s) Log Mine: %s - %s" % (self.args.project, self.args.source, datetime.now().strftime("%Y-%m-%d"))

        self.logger.info(subject)
        self.logger.info(body)

        message = body
        server = smtplib.SMTP('localhost')
        server.set_debuglevel(0)

        msg = MIMEMultipart('alternative')
        msg['Subject'] = subject
        msg['From'] = sender
        msg['To'] = recipient
        content = MIMEText(message, 'html')
        msg.attach(content)
        self.logger.debug("send mail from {} to {}".format(sender, recipient))
        server.sendmail(sender, [recipient], msg.as_string())
        server.quit()


if __name__ == "__main__":
    """
# query, parse and analyze logs using "artificial ignorance"
python -m boom.tools.logs.LogMine \
--log_filter "resource.type="container" resource.labels.cluster_name="qa-trinity" severity>=WARNING" \
--output_dir ./logmine \
--source qa_trinity \
--limit 1000 \
--range_in_minutes 30 \
--max_entry_count 20 \
--output_file qa_trinity_output.log    


python -m boom.tools.logs.LogMine \
--send_file true \
--output_dir ./logmine \
--output_file stopfile_airflow_dyna.txt
    """

    lm = LogMine()

File: logmine/logmine.go
package main

import (
	"flag"
	"fmt"
	"strings"
	"golang.org/x/net/context"
	"cloud.google.com/go/logging"
	"cloud.google.com/go/logging/logadmin"
	"google.golang.org/api/dataflow/v1b3"
	"google.golang.org/api/iterator"
	"log"
	"time"
	"os"
	"bufio"
	"runtime"
	"encoding/json"
	"gopkg.in/gomail.v2"
	"strconv"
	"golang.org/x/oauth2/google"
)

/*
go run logmine.go \
-logFilter='resource.type="container" resource.labels.cluster_name="qa-trinity" severity>=WARNING' \
-outputDir=./output \
-source=qa_trinity \
-maxEntryCount=20 \
-rangeInMinutes=10 \
-outputFile=qa_trinity_output.log

go run logmine.go \
-logFilter='resource.type=dataflow_step resource.labels.job_id=DATAFLOW_JOB_NAME' \
-outputDir=./output \
-source=qa-trinity-event-store \
-maxEntryCount=20 \
-rangeInMinutes=10 \
-outputFile=qa_trinity_event_store_output.log
 */

type StopFile struct {
	entry  string
	count  []int
}

type StopFiles []StopFile

// define a struct to pass around args since we don't have classes with member fields or functions
type logmine struct {
	filter  string
	minutes	int
	source  string
	project string
	count   int
	dir     string
	fname   string
	user    string
	password  string
}

func main() {
	logFilter := flag.String("logFilter", "resource.type=\"container\" resource.labels.cluster_name=\"qa-trinity\" severity>=WARNING", "log query filter")
	maxEntryCount := flag.Int("maxEntryCount", 20, "max per log entry to save to file")
	outputDir := flag.String("outputDir", "/tmp", "directory to store all files")
	outputFile := flag.String("outputFile", "/output.log", "output file name")
	rangeInMinutes := flag.Int("rangeInMinutes", 0, "past number if minutes to query")
	sender := flag.String("sender", "no-reply@nestlabs.com", "email sender")
	recipients := flag.String("recipients", "tlichtenberg@google.com", "email recipients")  // TODO figure out slice flags or comma-split this
	source := flag.String("source", "", "log source")
	project := flag.String("project", "nest-ds-dev", "gcp project")
	user := flag.String("user", "", "gcp project")
	password := flag.String("password", "", "gcp project")

	// got to flag.Parse() else defaults remain as values
	flag.Parse()

	// initialize the logmine struct with arg values
	mine := logmine{
					   filter: *logFilter,
					   minutes: *rangeInMinutes,
					   source: *source,
					   project: *project,
					   count: *maxEntryCount,
					   dir: *outputDir,
					   fname: *outputFile,
					   user: *user,
					   password: *password,
				   }

	// got to work with flag values AFTER parse
	recipientsList := strings.Split(*recipients, ",")
	fmt.Printf("sender = %v\n", *sender)
	// iterate through the slice using range
	for i := range recipientsList {
		fmt.Println(recipientsList[i])
	}

	ctx := context.Background()
	//dataflowService, err := dataflow.New(oauthHttpClient)
	adminClient, err := logadmin.NewClient(ctx, *project)
	if err != nil {
		log.Fatalf("Failed to create logadmin client: %v", err)
	}

	printMine(mine)
	entries, err := doQuery(adminClient, mine)
	newData, _ := doProcess(mine, entries)
	doSendMail(mine, *sender, recipientsList, newData)
}

func printMine(mine logmine) {
	fmt.Printf("maxEntryCount = %v\n", mine.count)
	fmt.Printf("outputDir = %v\n", mine.dir)
	fmt.Printf("outputFile = %v\n", mine.fname)
	fmt.Printf("rangeInMinutes = %v\n", mine.minutes)
	fmt.Printf("source = %v\n", mine.source)
	fmt.Printf("project = %v\n", mine.project)
	fmt.Printf("filter = %v\n", mine.filter)
}

func doQuery(adminClient *logadmin.Client, mine logmine) ([]*logging.Entry, error)  {

	var entries []*logging.Entry
	ctx := context.Background()

	// process filter
	filter := doFilterRevisions(mine)

	// [START list_log_entries]
	const name = "history"
	iter := adminClient.Entries(ctx,
		// Only get entries from the log-example log.
		logadmin.Filter(fmt.Sprintf("%s", filter)),
		// Get most recent entries first.
		logadmin.NewestFirst(),
	)

	// fetch the most recent maxEntryCount entries.
	for len(entries) < mine.count {
		entry, err := iter.Next()
		if err == iterator.Done {
			return entries, nil
		}
		if err != nil {
			return nil, err
		}
		entries = append(entries, entry)
	}

	log.Printf("Found %d entries.", len(entries))
	for _, entry := range entries {
		fmt.Printf("Entry: %6s @%s: %v\n\n",
			entry.Severity,
			entry.Timestamp.Format(time.RFC3339),
			entry.Payload)
	}
	// [END list_log_entries]

	return entries, nil
}

func doProcess(mine logmine, entries []*logging.Entry) ([]StopFile, []StopFile) {
	var newData = []StopFile{}
	var dataCounts = make(map[string] int)

	// read in existing stopfile, if any
	stopfile, err := readStopfile(mine)

	// read in the filter file data, if any
	filterFileData, _ := readFilterfile(mine)

	// count up all occurrences of each entry in the dataCounts map
	for i := range entries {
		payload := fmt.Sprintf("%s", entries[i].Payload)  // convert payload to a string
		//fmt.Printf("doProcess.payload: %s\n", payload)
		entry := doFilter(mine, payload, filterFileData)  // match against filter file data
		//fmt.Printf("doProcess.entry: %s\n\n", entry)
		if _, ok := dataCounts[entry]; ok {
			//fmt.Printf("incrementing dataCount for %s\n", entry)
			dataCounts[entry] += 1
		} else {
			//fmt.Printf("initializing dataCount for %s\n", entry)
			dataCounts[entry] = 1
		}
	}

	// if the dataCounts entry is in the stopfile, append its count
	// else make a new key in the stopfile for the entry with its count
	for k,v := range dataCounts {
		//fmt.Printf("k = %s, v = %d\n", k, v)
		var foundEntry = false
		for i := range stopfile {
			//fmt.Printf("stopfile entry: %s\n", stopfile[i].entry)
			a := strconv.QuoteToASCII(stopfile[i].entry)
			b := strconv.QuoteToASCII(k)
			if strings.Contains(a, b) {
				stopfile[i].count = append(stopfile[i].count, v)
				if len(stopfile[i].count) > mine.count {
					stopfile[i].count = stopfile[i].count[len(stopfile[i].count)-mine.count:]
				}
				foundEntry = true
				break
				}
		}
		if foundEntry == false {
			var c = make([]int,0)
			c = append(c, v)
			stopfile = append(stopfile, StopFile{entry: k, count: c})
			newData = append(newData, StopFile{entry: k, count: c})
		}
	}

	fmt.Printf("stopfile: %v\n\n", stopfile)
	err = writeStopfile(mine, stopfile)
	Check(err)

	fmt.Printf("newData: %v\n\n", newData)
	err = writeNewfile(mine, newData)
	Check(err)

	return newData, stopfile
}

func doFilter(mine logmine, logEntry string, filterFileData []string) string {
	for _, line :=  range filterFileData {
		line = strings.TrimSuffix(line, "\n")
		if strings.Contains(logEntry, line) {
			fmt.Printf("found filter line: %s\n", line)
			logEntry = line
			break
		}
	}
	return logEntry
}

func doFilterRevisions(mine logmine) string {
	newFilter := mine.filter
	if strings.Contains(mine.filter, "DATAFLOW_JOB_NAME") {
		newFilter = strings.Replace(mine.filter, "DATAFLOW_JOB_NAME", getDataflowJobByName(mine), 1)
	}

	if mine.minutes > 0 { // TODO verify this works. saw weird stuff on https://play.golang.org/
		minutes := time.Duration(mine.minutes)
		currentTime := time.Now()
		//fmt.Println("Current Time in String: ", currentTime.String())
		after := currentTime.Add(-minutes*time.Minute)
		newAfter := fmt.Sprintf(after.Format(time.RFC3339))
		newFilter = fmt.Sprintf("%s timestamp > \"%v\"", newFilter, newAfter)
		fmt.Println(newFilter)
	}

	return newFilter
}

func getDataflowJobByName(mine logmine) string {
	var jobId = "DATAFLOW_JOB_NOT_FOUND"
	ctx := context.Background()
	oauthHttpClient, err := google.DefaultClient(ctx,
		"https://www.googleapis.com/auth/devstorage.full_control")
	if err != nil {
		log.Fatal(err)
	}
	dataflowService, err := dataflow.New(oauthHttpClient)
	f := dataflowService.Projects.Jobs.List(mine.project)
	f.Filter("ACTIVE")
	jobs, err := f.Do()
	for _, job := range jobs.Jobs {
		//fmt.Printf("Name: %s, Id: %s\n", job.Name, job.Id)
		if strings.Contains(job.Name, mine.source) {
			jobId = job.Id
			fmt.Printf("found dataflow job name: %s, returning job id: %s\n", job.Name, job.Id)
			break
		}
	}
	return jobId
}

func writeTextfile(outfile string, stopfiles []StopFile) error {
	f, err := os.Create(outfile)
	Check(err)
	defer f.Close()
	for _, v := range stopfiles {
		s := fmt.Sprintf("%s:%v\n", v.entry, v.count)
		_, err = f.WriteString(s)
		f.Sync()
	}
	return err
}

func readTextfile(infile string) ([]string, error) {
	var lines []string
	file, err := os.Open(infile)
	if err != nil {
		return lines, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func writeStopfile(mine logmine, stopfile []StopFile) error {
	outfile := fmt.Sprintf("%s/stopfile_%s.txt", mine.dir, mine.source)
	err := writeTextfile(outfile, stopfile)
	return err
}

func readStopfile(mine logmine) ([]StopFile, error) {
	infile := fmt.Sprintf("%s/stopfile_%s.txt", mine.dir, mine.source)
	stopfiles := []StopFile{}
	lines, err := readTextfile(infile)
	for i :=  range lines {
		line := strings.Split(lines[i],":")
		var ints []int
		err = json.Unmarshal([]byte(line[1]), &ints)
		stop := StopFile{entry: line[0], count: ints}
		stopfiles = append(stopfiles, stop)
	}
	return stopfiles, err
}

func readFilterfile(mine logmine) ([]string, error) {
	infile := fmt.Sprintf("%s/filterfile_%s.txt", mine.dir, mine.source)
	lines, err := readTextfile(infile)
	return lines, err
}

func writeNewfile(mine logmine, newfile []StopFile) error {
	outfile := fmt.Sprintf("%s/newfile_%s.txt", mine.dir, mine.source)
	err := writeTextfile(outfile, newfile)
	return err
}

func Check(e error) {
	if e != nil {
		_, file, line, _ := runtime.Caller(1)
		fmt.Println(line, "\t", file, "\n", e)
		os.Exit(1)
	}
}

func doSendMail(mine logmine, sender string, recipients []string,  data []StopFile)  {
	//auth := smtp.PlainAuth("", mine.user, mine.password, "localhost")
	htmlData := "<body><table border=2><tr><td> New Entries </td><td> Entry Count </td></tr>"
	for _, v := range data {
		htmlDataPart := fmt.Sprintf("<tr><td>%s</td><td>%v</td></tr>", v.entry, v.count)
		htmlData += htmlDataPart
	}
	htmlData += "</table></body>"
	//mime := "MIME-version: 1.0;\nContent-Type: text/html; charset=\"UTF-8\";\n\n";
	subj := fmt.Sprintf("(%s) Log Mine [Go]: %s - %s", mine.project, mine.source, time.Now())
	// msg := []byte(subj + mime + htmlData)

	for _, recipient := range recipients {
		m := gomail.NewMessage()
		m.SetHeader("From", sender)
		m.SetHeader("To", recipient)
		m.SetHeader("Subject", subj)
		m.SetBody("text/html", htmlData)

		// Send the email
		d := gomail.NewPlainDialer("localhost", 25, mine.user, mine.password)
		if err := d.DialAndSend(m); err != nil {
			fmt.Println(err)
		}
	}
	/*
	err := smtp.SendMail("localhost:25", auth, sender, recipient, msg)
	if err != nil {
		fmt.Printf("sendmail error: %s", err)
	}
	*/
}